Issue 1228

UpworkBartkoski opened the pull request:
### Details

When hitting on the arrow down, the focus moves into `ReportActionCompose` TextInput and it is active to input.
That's why this issue happened. `ReportActionCompose` TextInput should be disabled.
Actually, `ReportActionCompose` is not needed to be editable when the sidebar is shown on the phone browser.

### Fixed Issues
#1228 

### Tests

1. On Search box Add one of your contact
2. Arrow down in keyboard will be disabled

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
![Screen Shot 2021-01-19 at 12 03 52 PM](https://user-images.githubusercontent.com/74012504/105075258-21beca00-5a4f-11eb-8bfc-afef84d912b2.png)

#### Web
![Screen Shot 2021-01-19 at 12 09 28 PM](https://user-images.githubusercontent.com/74012504/105075374-4e72e180-5a4f-11eb-9922-7ffdcb0a1d45.png)

#### Mobile Web
![Screen Shot 2021-01-19 at 12 10 48 PM](https://user-images.githubusercontent.com/74012504/105075451-63e80b80-5a4f-11eb-961b-69c1259e33a2.png)

#### Desktop
![Screen Shot 2021-01-19 at 12 18 48 PM](https://user-images.githubusercontent.com/74012504/105076337-8a5a7680-5a50-11eb-8236-c4d54d0ff150.png)

#### iOS
![Screen Shot 2021-01-19 at 12 24 51 PM](https://user-images.githubusercontent.com/74012504/105076940-5fbced80-5a51-11eb-8496-d331256a69b6.png)


#### Android

UpworkBartkoski committed:
fix: #1228

UpworkBartkoski committed:
Merge branch 'master' of https://github.com/Expensify/Expensify.cash into issue-1228

Jag96 reviewed (changes requested):
It looks like on web if you start off with a small width window, open the side menu, type in the search box, and then make the entire window larger you can end up with a full-screen app where the compose TextInput is disabled. Can you update your solution to prevent this regression?

Gif: http://g.recordit.co/EQmL6pEIXI.gif

There have also been some big changes to the code recently, so please merge master into your branch before testing your new fix to ensure that your fix for the issue includes those changes.

UpworkBartkoski committed:
Merge branch 'master' of https://github.com/Expensify/Expensify.cash into issue-1228

roryabraham commented on src/components/TextInputFocusable/index.js:
```diff
@@ -35,6 +35,9 @@ const propTypes = {
     onDrop: PropTypes.func,
+
+    // If the chatswitcher is shown in iOS safari browser, the input should be disabled
+    isDisable: PropTypes.bool,
```
Let's call this `isDisabled` instead of `isDisable`

UpworkBartkoski replied:
Updated.

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -15,6 +21,8 @@ import {addAction, saveReportComment, broadcastUserIsTyping} from '../../../libs
 import ReportTypingIndicator from './ReportTypingIndicator';
 import AttachmentModal from '../../../components/AttachmentModal';
 
+const windowSize = Dimensions.get('window');
```
Can you use the [`useWindowDimensions` hook](https://reactnative.dev/docs/usewindowdimensions) instead of `Dimensions.get('window')`? 

By doing something like this: 

```javascript
import { useWindowDimensions } from 'react-native';
const windowWidth = useWindowDimensions().width;
```

the value of `windowWidth` will automatically be updated when the window resizes. In your current solution, you measure the window width once before the component renders, so it doesn't get updated when you resize the window.

UpworkBartkoski replied:
`ReportActionCompose` is not a functional component.
How can use `useWindowDimensions` hook inside in a class component directly?

roryabraham replied:
You're right, my mistake

UpworkBartkoski committed:
minor change

UpworkBartkoski committed:
change isDisable to isDisabled

UpworkBartkoski commented:
> It looks like on web if you start off with a small width window, open the side menu, type in the search box, and then make the entire window larger you can end up with a full-screen app where the compose TextInput is disabled. Can you update your solution to prevent this regression?
> 
> Gif: http://g.recordit.co/EQmL6pEIXI.gif
> 
> There have also been some big changes to the code recently, so please merge master into your branch before testing your new fix to ensure that your fix for the issue includes those changes.

Updated

roryabraham reviewed (approved):
Thanks for making those updates. LGTM 👍

roryabraham reviewed (changes requested):
Actually, after discussing in the issue, I agree that a different solution might be better. Sorry for the inconvenience, but could we actually approach this differently, and do something like this:

If the ChatSwitcherSearchForm is in focus, and you hit the down arrow, then it will switch to the ReportActionCompose (like the default behavior), but also hide the sidebar using [this function](https://github.com/Expensify/Expensify.cash/blob/7afeb45b8e7833c7c9e67f942318210f3ac5a1ec/src/libs/actions/Sidebar.js#L7-L9).

UpworkBartkoski commented:
> Actually, after discussing in the issue, I agree that a different solution might be better. Sorry for the inconvenience, but could we actually approach this differently, and do something like this:
> 
> If the ChatSwitcherSearchForm is in focus, and you hit the down arrow, then it will switch to the ReportActionCompose (like the default behavior), but also hide the sidebar using [this function](https://github.com/Expensify/Expensify.cash/blob/7afeb45b8e7833c7c9e67f942318210f3ac5a1ec/src/libs/actions/Sidebar.js#L7-L9).

Yes, then it will be easier.

roryabraham commented:
Cool, can you also make sure that the inverse works okay: if the ReportActionCompose has focus, then using the up arrow to focus the `ChatSwitcherSearchForm` will show the sidebar?

UpworkBartkoski commented:
> Cool, can you also make sure that the inverse works okay: if the ReportActionCompose has focus, then using the up arrow to focus the `ChatSwitcherSearchForm` will show the sidebar?

I am not sure if I can get listener of arrow icon in mobile browser.
Do you have any idea?

roryabraham commented:
> I am not sure if I can get listener of arrow icon in mobile browser.

Can you just listen for focus events on the respective text inputs? I.e: if the ReportActionCompose gets focused make sure the sidebar is not visible, and if the ChatSwitcherSearchForm gets focused make sure the sidebar _is_ visible. Ideally we wouldn't actually be worrying about the arrow buttons themselves and rather just let them operate as they should

UpworkBartkoski commented:
> > I am not sure if I can get listener of arrow icon in mobile browser.
> 
> Can you just listen for focus events on the respective text inputs? I.e: if the ReportActionCompose gets focused make sure the sidebar is not visible, and if the ChatSwitcherSearchForm gets focused make sure the sidebar _is_ visible. Ideally we wouldn't actually be worrying about the arrow buttons themselves and rather just let them operate as they should

Yes, I did like that.
But, animation or other ones doesn't work properly.
If you want to see the result, I will push the change.

roryabraham commented:
Sure, feel free to push the change and describe the problem

UpworkBartkoski commented:
> Sure, feel free to push the change and describe the problem

Updated.

Jag96 commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -103,7 +106,7 @@ class ChatSwitcherView extends React.Component {
                 && !this.props.isSidebarAnimating
         ) {
-            this.textInput.focus();
+            // this.textInput.focus();
```
Is there a reason we're removing this focus check now? If you did intend to remove this, please remove the code instead of just commenting it out.

UpworkBartkoski replied:
Because of it, the focus can't stay on `ReportActionCompose`.
Also, that condition doesn't work except when a keyboard event fires.

UpworkBartkoski commented:
Waiting for review...

Jag96 reviewed (dismissed):
@UpworkBartkoski sorry about the back and forth here, I've commented on [the issue](https://github.com/Expensify/Expensify.cash/issues/1228#issuecomment-769451517) to confirm the proposed solution here (It looks like you had a solution for this before [this comment](https://github.com/Expensify/Expensify.cash/pull/1268#pullrequestreview-573496972)). Please update the PR and PR description and we will review again.

UpworkBartkoski commented:
> @UpworkBartkoski sorry about the back and forth here, I've commented on [the issue](https://github.com/Expensify/Expensify.cash/issues/1228#issuecomment-769451517) to confirm the proposed solution here (It looks like you had a solution for this before [this comment](https://github.com/Expensify/Expensify.cash/pull/1268#pullrequestreview-573496972)). Please update the PR and PR description and we will review again.

@Jag96
So should I go back to the first solution I did?
3cf5b69

Jag96 commented:
@UpworkBartkoski yes, if https://github.com/Expensify/Expensify.cash/commit/3cf5b69093a15a6f84a6f29a4328c67b33754f6a prevents the arrows from being tappable and doesn't cause any other regressions that should be fine. Once it's updated we'll review it again.

UpworkBartkoski committed:
fix merge conflict

UpworkBartkoski commented:
> @UpworkBartkoski yes, if [3cf5b69](https://github.com/Expensify/Expensify.cash/commit/3cf5b69093a15a6f84a6f29a4328c67b33754f6a) prevents the arrows from being tappable and doesn't cause any other regressions that should be fine. Once it's updated we'll review it again.

Updated

roryabraham commented on src/components/TextInputFocusable/index.js:
```diff
@@ -35,6 +35,9 @@ const propTypes = {
     // Callback to fire when a file is dropped on the text input
     onDrop: PropTypes.func,
+
+    // If the chatswitcher is shown in iOS safari browser, the input should be disabled
```
```suggestion
    // Whether or not this TextInput is disabled.
```

UpworkBartkoski replied:
Changed.

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -134,7 +155,21 @@ class ReportActionCompose extends React.Component {
+    }
+
     render() {
+        const inputDisable = this.props.isSidebarShown
```
I would add a comment here:

// We want to make sure to disable on small screens because in iOS safari the keyboard up/down buttons will focus this from the chat switcher. 
// https://github.com/Expensify/Expensify.cash/issues/1228

UpworkBartkoski replied:
Replaced comment.

roryabraham reviewed (changes requested):
Since you originally created this PR, we added a HOC called [withWindowDimensions](https://github.com/Expensify/Expensify.cash/blob/43512e020f9933da5cd328c468705f8fa60d85d5/src/components/withWindowDimensions.js). Could you please:

1) Update that HOC:
    a) Add `isSmallScreenWidth` to `windowDimensionsPropTypes`
    b) Pass `isSmallScreenWidth={this.state.windowDimensions.width <= variables.mobileResponsiveWidthBreakpoint}` to the wrapped component as an additional prop.
2) Then in `ReportActionCompose`:
    a) Wrap ReportActionCompose with the `withWindowDimensions` HOC, and add `...windowDimensionsPropTypes` to the props validation.
    b) Take `isSmallScreenWidth` out of the state and get rid of any calls to `Dimensions` along with your `componentDidMount`, `componentWillUnmount`, and `disableInputBasedOnDimensions`
    c) In render, set `const inputDisable = this.props.isSidebarShown && this.props.isSmallScreenWidth`

UpworkBartkoski committed:
Merge branch 'master' into issue-1228

UpworkBartkoski committed:
updates for request changes

UpworkBartkoski commented:
> Since you originally created this PR, we added a HOC called [withWindowDimensions](https://github.com/Expensify/Expensify.cash/blob/43512e020f9933da5cd328c468705f8fa60d85d5/src/components/withWindowDimensions.js). Could you please:
> 
> 1. Update that HOC:
>    a) Add `isSmallScreenWidth` to `windowDimensionsPropTypes`
>    b) Pass `isSmallScreenWidth={this.state.windowDimensions.width <= variables.mobileResponsiveWidthBreakpoint}` to the wrapped component as an additional prop.
> 2. Then in `ReportActionCompose`:
>    a) Wrap ReportActionCompose with the `withWindowDimensions` HOC, and add `...windowDimensionsPropTypes` to the props validation.
>    b) Take `isSmallScreenWidth` out of the state and get rid of any calls to `Dimensions` along with your `componentDidMount`, `componentWillUnmount`, and `disableInputBasedOnDimensions`
>    c) In render, set `const inputDisable = this.props.isSidebarShown && this.props.isSmallScreenWidth`

Updated.

roryabraham commented on src/components/withWindowDimensions.js:
```diff
@@ -11,6 +12,8 @@ const windowDimensionsPropTypes = {
         // Height of the window
         height: PropTypes.number.isRequired,
+
+        isSmallScreenWidth: PropTypes.bool,
```
This should be outside of the `windowDimensions` prop. Like this:

```javascript
const windowDimensionsPropTypes = {
    // via withWindowDimensions
    windowDimensions: PropTypes.shape({
        // Width of the window
        width: PropTypes.number.isRequired,

        // Height of the window
        height: PropTypes.number.isRequired,
    }).isRequired,
    
    // Is the window width narrow, like on a mobile device?
    isSmallScreenWidth: PropTypes.bool.isRequired,
};
```

Also note that `isSmallScreenWidth` should be required. Our style guide mandates that any optional props should always have a default. In this case, the prop will always be provided by this HOC, so it's better to just make it a required prop.

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -14,6 +14,7 @@ import AttachmentPicker from '../../../components/AttachmentPicker';
 import {addAction, saveReportComment, broadcastUserIsTyping} from '../../../libs/actions/Report';
 import ReportTypingIndicator from './ReportTypingIndicator';
 import AttachmentModal from '../../../components/AttachmentModal';
+import withWindowDimensions from '../../../components/withWindowDimensions';
```
```suggestion
import withWindowDimensions, {windowDimensionsPropTypes} from '../../../components/withWindowDimensions';
```

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -24,10 +25,16 @@ const propTypes = {
+
+    isSidebarShown: PropTypes.bool,
+
+    isSmallScreenWidth: PropTypes.bool,
```
```suggestion
    windowDimensionsPropTypes.isSmallScreenWidth,
```

roryabraham replied:
Or just `...windowDimensionsPropTypes`

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -24,10 +25,16 @@ const propTypes = {
     // The ID of the report actions will be created for
     reportID: PropTypes.number.isRequired,
+
+    isSidebarShown: PropTypes.bool,
```
```suggestion
    isSidebarShown: PropTypes.bool.isRequired,
```

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -24,10 +25,16 @@ const propTypes = {
 const defaultProps = {
     comment: '',
+    isSidebarShown: true,
+    isSmallScreenWidth: false,
```
`isSmallScreenWidth` should be required.

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -24,10 +25,16 @@ const propTypes = {
 
 const defaultProps = {
     comment: '',
+    isSidebarShown: true,
```
I think in the ReportActionCompose, we should assume that `isSidebarShown` is false, because on mobile under normal circumstances this component is only visible if the sidebar is not.

```suggestion
    isSidebarShown: false,
```

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -135,6 +143,10 @@ class ReportActionCompose extends React.Component {
 
     render() {
+        // We want to make sure to disable on small screens because in iOS safari the keyboard up/down buttons will
+        // focus this from the chat switcher. #1228
```
Sorry, that was just github collapsing the link to the issue, but this should be the full URL.

```suggestion
        // focus this from the chat switcher. https://github.com/Expensify/Expensify.cash/issues/1228
```

UpworkBartkoski committed:
fix code format

roryabraham reviewed (approved):
Okay, looks great and tests well. Thanks!
